openoa/analysis/wake_losses.py

```diff
@@ -0,0 +1,1584 @@
+        turbine_wake_losses_lt = len(self.turbine_ids) * [0.0]
+        for i, t in enumerate(self.turbine_ids):
+            # determine ideal turbine energy as sum of the power produced by the turbine when it is derated and the
+            # mean power produced by all freestream turbines when the turbine is operating normally
+
+            df_1hr_bin[("energy_avg", t)] = df_1hr_bin["freq"] * df_1hr_bin[("power", t)]
+            df_1hr_bin[("potential_turbine_energy", t)] = (
+                df_1hr_bin["freq"] * df_1hr_bin[("potential_turbine_power", t)]
+            )
+
+            turbine_wake_losses_lt[i] = 1 - (
+                df_1hr_bin[("energy_avg", t)].sum()
+                / df_1hr_bin[("potential_turbine_energy", t)].sum()
+            )
+
+        # Save long-term corrected plant and turbine-level wake losses binned by wind direction
+        df_1hr_wd_bin = df_1hr_bin.groupby(level=[0]).sum()
+
+        wake_losses_lt_wd = (
+            df_1hr_wd_bin["actual_plant_energy"] / df_1hr_wd_bin["potential_plant_energy"]
+        ).values
+
+        energy_lt_wd = (
+            df_1hr_wd_bin["actual_plant_energy"].values / df_1hr_wd_bin["actual_plant_energy"].sum()
+        )
+
+        turbine_wake_losses_lt_wd = np.empty(
+            [len(self.turbine_ids), int(360.0 / self._wd_bin_width_LT_corr)]
+        )
+        for i, t in enumerate(self.turbine_ids):
+            turbine_wake_losses_lt_wd[i, :] = (
+                df_1hr_wd_bin[("energy_avg", t)] / df_1hr_wd_bin[("potential_turbine_energy", t)]
+            ).values
+
+        # Save long-term corrected plant and turbine-level wake losses binned by wind speed
+        df_1hr_ws_bin = df_1hr_bin.groupby(level=[1]).sum()
+
+        # reindex to fill in missing wind speed bins
+        index = np.arange(0.0, 31.0, self._ws_bin_width_LT_corr).tolist()
+        df_1hr_ws_bin = df_1hr_ws_bin.reindex(index)
+
+        wake_losses_lt_ws = (
+            df_1hr_ws_bin["actual_plant_energy"] / df_1hr_ws_bin["potential_plant_energy"]
+        ).values
+
+        energy_lt_ws = (
+            df_1hr_ws_bin["actual_plant_energy"].values / df_1hr_ws_bin["actual_plant_energy"].sum()
+        )
+
+        turbine_wake_losses_lt_ws = np.empty(
+            [len(self.turbine_ids), int(30.0 / self._ws_bin_width_LT_corr) + 1]
+        )
+        for i, t in enumerate(self.turbine_ids):
+            turbine_wake_losses_lt_ws[i, :] = (
+                df_1hr_ws_bin[("energy_avg", t)] / df_1hr_ws_bin[("potential_turbine_energy", t)]
+            ).values
+
+        return (
+            wake_losses_lt,
+            turbine_wake_losses_lt,
+            wake_losses_lt_wd,
+            turbine_wake_losses_lt_wd,
+            energy_lt_wd,
+            wake_losses_por_ws,
+            turbine_wake_losses_por_ws,
+            energy_por_ws,
+            wake_losses_lt_ws,
+            turbine_wake_losses_lt_ws,
+            energy_lt_ws,
+        )
+
+    def plot_wake_losses_by_wind_direction(
```

Comment:
I made the API similar to the other analysis class plotting functions by allowing customized inputs. I also moved the main plotting function to the `plot` utils module, similar to other analysis classes, to simplify the wake loss plotting functions in `wake_losses`.

Reply:
Nice work on getting the two plotting routines here reduced to a single routine in the plot module so that each customized routine relies on a different data view going into the generalized routine!